### ISSUE TITLE
start navigation can fail to correctly load next view controller

### ISSUE THREAD

**Issue by sawarren** [tallygo/TallyGoKit#14]
TallyGoKit version: 2.0.3 
iOS version: 11.3.1
Hardware model: iPhone 6S Plus

I initialise a new instance of TGPreviewViewController with a default TGTurnByTurnConfiguration, which
matches the documentation: https://tallygo.com/docs/ios/get-nav-in-app

Naturally, the origin and destinations are dynamic based on user location and input.

Upon presenting the view controller and tapping 'Start Navigation' the view controller can fail to load, the result being the below image.

![issue](https://user-images.githubusercontent.com/18529344/42730681-6b15f80e-87f2-11e8-8ff9-662c3623988f.jpg)

**Comment by mtriplett** [comment]
Can you share your code with us? Something is definitely wonky, but we are unable to reproduce this on our end.

**Comment by mtriplett** [comment]
We have just release iOS SDK version 2.0.4 which should resolve this issue.

https://github.com/tallygo/TallyGoKit/releases/tag/2.0.4

You should be able to get the latest by running `pod update` (if using CocoaPods). It should update to 2.0.4. If CocoaPods is being stubborn and not seeing 2.0.4 yet, run `pod repo update` and then `pod update` again.

Can you let us know if this fixes the issue for you? Much appreciated!

**Comment by sawarren** [comment]
Sorry for the delay in response.

I have integrated the new version of TallyGoKit and the issue does not seem to be resolved.

I have attached a screenshot of the code, but as I said in the original report it is basically the code that is in the documentation

![screen shot 2018-07-21 at 19 46 50](https://user-images.githubusercontent.com/18529344/43039168-ab9a14fa-8d1f-11e8-89dd-d30dfbcf4218.png)

Thanks

**Comment by sawarren** [comment]
Have you had any luck debugging this issue?

this state, once achieved, can be resolved by pressing the pause button on the lower left and, upon unpausing, the screen UI loads as expected.

**Comment by mtriplett** [comment]
We haven't been able to reproduce the issue. I'd suggest you try the basic navigation example (https://www.tallygo.com/docs/ios/get-nav-in-app) and confirm that it works. If that doesn't work, then we know there's a bigger issue at play here. If that basic example DOES work, then I'd recommend creating a super basic app with just the navigation example but not using the simulator so you can see if that works or not.

There's clearly something interfering with the navigation from starting up properly. It's just a matter of blocking and tackling to get it narrowed down.

Let us know what you find out.

**Comment by sawarren** [comment]
Hi Matt,

I copied and pasted the code in the link (even though my code is the same)
and the issue is the same when using that code. This was tested on an
iphone 6s plus, not the simulator

I'd like to clarify the issue, in case I haven't explained it well enough.

When you press the 'start navigation' button on the
TGPreviewViewController, the screen it pushes is not loaded correctly.

when the screen is paused and unpaused, this issue is resolved.

My suggestion would be investigating the unpausing code, as whatever logic takes
place there resolves the issue.

Again, all my code is doing is presenting the preview controller, which
seems to work as intended.

it is the loading of the next screen that is the issue.

if needed, you can ask hayden to retrieve the video which i sent him that
shows the issue clearly.

**Comment by dmdeller** [comment]
Hi Steven,

Sorry you're continuing to experience this.

Do I understand correctly that the problem occurs when you are not using the simulator, but does not occur when you are using the simulator?

Would you be able to provide us with sample origin and destination coordinates that result in the issue?

Also, note that our coverage area for turn-by-turn navigation currently only includes the continental United States. For developers outside of that area, it should not be an issue when you use simulated coordinates inside that area. However, even if you are using coordinates (real or simulated) outside of that area, it is supposed to fail gracefully with an error message, rather than result in a broken screen as in your screen shots, so we would like to try and track this down regardless.

Thanks!

**Comment by dmdeller** [comment]
Also, to clarify: when we say 'simulator' we are generally referring to our driving simulator, a feature of the TallyGo SDK, rather than Apple's iOS Simulator. Our driving simulator can be used either in the iOS Simulator or on a real device. This is determined by the presence our absence of the call to `TallyGo.simulatedCoordinate`.

**Comment by sawarren** [comment]
Hi there,

Firstly, I am based in Edinburgh, so in order to test the application, I am using gpx files to simulate the phones location in Xcode. The app in question is a parking app that operates in LA.

I wanted to make sure this was an SDK issue and not a project issue, so I implemented the exact code in the docs, including `TallyGo.simulatedCoordinate`, **_in a different project_**. And it worked, which was confusing. 

I will say, as a side note, that the screen did look blank for a few seconds (like the image above). As such, I may have closed the application prematurely in my previous test with `TallyGo.simulatedCoordinate`. I would recommend adding a `UIActivityIndicatorView` to avoid this.

Anyway, I then added a gpx coordinate to simulate the phones location in Los Angeles, incase that was the issue. It still worked. Again, confusing.

I then wanted to try it without `TallyGo.simulatedCoordinate`. However, I did not have `NSLocationAlwaysUsageDescription` or `NSLocationWhenInUseUsageDescription` in my plist, so I got errors about that.

Once I added those, I checked if it worked.
With `TallyGo.simulatedCoordinate`, it did. but without it, it did not.

To be thorough, I also then tried the original implementation, in the original project with a `TallyGo.simulatedCoordinate`. This worked

This leads me to the conclusion that your implementation waits for a location update before updating the UI? 

As for a sample origin and destination that fails, the ones in your docs fail:

```
     // Get these coordinates from your app, these are just a sample
     let origin = CLLocationCoordinate2D(latitude: 34.101558, longitude: -118.340944) // Grauman's Chinese Theatre
     let destination = CLLocationCoordinate2D(latitude: 34.011441, longitude: -118.494932) // Santa Monica Pier
```

As a test case, I would suggest the following parameters:
- no `TallyGo.simulatedCoordinate` value
- leave the device flat on the table (dont move it)
- present a preview controller with the above origin and destination
- tap start navigation

TL;DR If I provide a `TallyGo.simulatedCoordinate` it works as intended, if i dont, it results in the above issue. this leads me to the conclusion that you are waiting for a location update to load your UI.

**Comment by mtriplett** [comment]
Very helpful info, Steven. Let us confer on this end and see what we can come up with. Stay tuned.

**Comment by dmdeller** [comment]
Hi Steven,

We received the GPX file you sent - thank you for that. Here are our findings so far.

- We were able to reproduce the issue with a GPX file with only one location in it.
- We were not able to reproduce the issue using a real device, flat and unmoving on a tabletop, using real location data.
- We were not able to reproduce the issue using the iOS Simulator, with a simulated (but static, unchanging) location, configured using Simulator.app’s menu items at Debug > Location > Apple.
- We were not able to reproduce the issue using TallyGoKit’s `simulatedCoordinate` option.

It seems you are correct that the UI does not update until it receives a location update. The reason that pausing and resuming fixes it seems to be because that stops and starts our `CLLocationManagerDelegate`, which seems to result in Xcode replaying the GPX file from the beginning. However, when using either real location data or the iOS Simulator’s simulated location, the operating system provides location updates continuously (even if the coordinates have not changed), apparently making this a non-issue in those cases. As far as we are aware, the issue only seems to manifest when using a GPX file with a single location in it.

We thank you for your report, and we hope to fix this issue at some point in the future, but as of now we do not have an ETA. Until then, for your testing needs, I would suggest using one of the other above techniques for simulating coordinates. Please let us know if there is anything else we can do to help.

**Comment by sawarren** [comment]
Alright. Thanks for notifying me. 

Happy developing.
